### PR TITLE
feat(github): Fetch issues during repo initialization

### LINKED ISSUE #27641: Permanent caching for GitHub issues and PRs
### Describe the proposed change(s).

It seems like we have solution for reliable caching:
- Use [`/repos/{owner}/{repo}/issues` endpoint](https://docs.github.com/en/rest/issues/issues?apiVersion=2022-11-28#list-repository-issues)
- Use `since` parameter, which works similarly to `If-Modified-Since` header
- Leverage the fact that empty result set always has same `ETag` value (seems like it's `89d4eb0a983ee9362287f95c869f0afd61584ac481494969bcaedc98e33feb50` across all the repos)
  - Although it's undocumented feature, probably we could rely on it (otherwise, it's problematic to combine `since` query param with `If-None-Match: <etag>` such that we ever receive `304`)
  - As an alternative, we could sync to the latest date until the result is empty, store empty result ETag, and perform the next sync only for 200 status: it's one of more complicated solutions I can think of, but maybe it's safer

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Modify `initRepo` method to fetch 5 last changed issues
- Detect if those issues are the only changed ones since the latest cache sync
- If so, then prevent new fetches during the run
- Additionally, add modified issue data after any of modification performed by Renovate

## Context

- Closes: #27641

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
- Blocked by #27782 
- Blocked by #27784

File: lib/modules/platform/github/issue.ts [hunk 1]
```diff
@@ -8,21 +9,132 @@ const GithubIssueBase = z.object({
...
+    if (!this.issuesToReconcile) {
+      return cacheData;
+    }
+
+    let isReconciled = false;
+
+    for (const issue of this.issuesToReconcile) {
+      const cachedIssue = cacheData[issue.number];
+
+      // If we reached the the item which is already in the cache,
```
Comment (viceice) [inline lib/modules/platform/github/issue.ts:111]:
```suggestion
      // If we reached the item which is already in the cache,
```

File: lib/modules/platform/github/index.ts [hunk 2]
```diff
@@ -555,6 +556,11 @@ export async function initRepo({
     config.autoMergeAllowed = repo.autoMergeAllowed;
     config.hasIssuesEnabled = repo.hasIssuesEnabled;
     config.hasVulnerabilityAlertsEnabled = repo.hasVulnerabilityAlertsEnabled;
+
+    const recentIssues = Issue.array()
```
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/modules/platform/github/index.ts:560]: Don't we have this problem anyway?

[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/platform/github/index.ts:560]: @zharinov Do we do a full refresh when this is empty? Can you add a test where the result is empty?

File: lib/modules/platform/github/issue.ts [hunk 3]
```diff
@@ -8,21 +9,132 @@ const GithubIssueBase = z.object({
...
+  }
+
+  static updateIssue(issue: GithubIssue): void {
+    const cacheData = this.data;
+    if (cacheData) {
+      cacheData[issue.number] = issue;
+    }
+  }
+
+  private static issuesToReconcile: GithubIssue[] | undefined;
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/modules/platform/github/issue.ts:84]: UPD: this attempt wasn't successful

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/modules/platform/github/issue.ts]: I believe that the repository "fingerprint" concept was a defensive measure aimed to reduce chance of accidental data leakage, right? If so then maybe we could still intentionally work around it for the initial query because if the fingerprint is wrong then we can act accordingly immediately after?

**Review by viceice — DISMISSED** [review]

**Comment by rarkins** [comment]
What about the repo fingerprint problem - should we fix that first?

**Comment by zharinov** [comment]
It's not easily fixable. It should be done in separate PR, and now I'm not sure if it's worth the struggle. GraphQL could not be cached anyway, and if we switch from GraphQL to the REST, it means we'll lose this issue fetch optimization and will have to fetch issues each run.

**Comment by rarkins** [comment]
How does it work now? It queries last 5 updated issues during initRepo(), but at that time the repository cache isn't initialized so we can't reconcile them.

So does it wait to do reconciliation the first time getIssues() is called?

**Comment by zharinov** [comment]
Oh, I've forgot that memory cache is probably also isn't available at the `initRepo()` phase, but this is easily fixable at least.

Anyway, the approach is to store the items obtained in `initRepo()` call, and reconcile them lazily the first time we need the cached items (at the moment when repo cache is already initialized).

**Review by rarkins — CHANGES_REQUESTED** [review]
Please add a log message which confirms if Issues are fully sync'd (usual case) or if for some reason we still need to do the graphql issues query to fetch more.

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]
